app/models/user.rb

```diff
@@ -565,6 +565,37 @@ def reactions_to
     Reaction.for_user(self)
   end
 
+  def latest_activity
+    return unless registered == true
+
+    latest_activity = last_comment_at || last_article_at || latest_article_updated_at ||
+      last_reacted_at || profile_updated_at || last_moderation_notification
+
+    if latest_activity.strftime("%Y-%m-%d") == Time.zone.today.strftime("%Y-%m-%d")
+      "Today, #{latest_activity.strftime('%d %b')}"
+    elsif latest_activity.strftime("%Y-%m-%d") == Date.yesterday.strftime("%Y-%m-%d")
+      "Yesterday, #{latest_activity.strftime('%d %b')}"
+    else
+      latest_activity.strftime("%d %b, %Y")
+    end
+  end
+
+  def current_role
+    if admin?
+      "Admin"
+    elsif super_admin?
+      "Super Admin"
+    elsif tech_admin?
+      "Tech Admin"
+    elsif tag_moderator?
+      "Moderator"
+    elsif has_role?(:single_resource_admin, :any)
+      "Resource Admin"
+    else
+      "Regular Member"
+    end
+  end
```

Comment:
I apologize in advance for the wall of text.  My intention is to share my context and also look to join in the conversations around this particular feature.  And offer any help to move forward with this.

I’m coming from a place where I’ve been wrestling with AuthN/AuthZ work for awhile and was a bit surprised to see the `User#current_role` method.  The surprise is because each user can have multiple roles, and the roles aren’t exactly hierarchical.

My question is if this is something in the design spec saying “we only want to render one role?”  It’s not quite a blocker, but there are some pitfalls I want to make sure to help guide you through.

In the specific implementation, I’m reading the code such that if someone has both the “Super Admin” and “Admin” role that they will show up as an “Admin”.

I’m also wondering about considerations for handling a few other relevant roles: `:trusted` and `:support_admin` in particular (see <https://github.com/forem/forem/blob/main/app/policies/authorizer.rb>).

A refactor that I’d like to see is moving the `current_role` method to the `./app/policies/authorizer.rb` and add a `delegate` method on the user (see <https://github.com/forem/forem/blob/9f445304915a5256c3cce6dc3609941ba23dde70/app/models/user.rb#L395-L418>).  The User model already has a lot of methods, so moving that to the Authorizer would help keep that `current_role` close to the functions it calls.

Another technical consideration is that the code may end up generating multiple queries per; This is something that will require a potentially different approach.  For each user in the search result, we would be potentially querying the role table multiple times.  This alone isn’t a blocker, but one that we want to record as a follow-up issue.  It’s possible a simple `includes(:roles)` will suffice, but I’m not optimistic.

All of this is to say, the roles of the code-base are tricky.  I know we want to read them as hierarchical, but as I’ve looked through the code, they aren’t quite.  There’s almost a hierarchy but in some cases things might leak through; (I found a case where a `tech_admin` and a `super_admin` can do something but not an `admin`).  There are orthogonal roles as well. An `admin` is not, by default a `moderator`; but a person could have both `admin` and `moderator` roles.

Reply:
> This makes sense, we'll record this as a follow up issue. @juliannatetreault will you please add this.
Absolutely, @Ridhwana! A follow-up issue sounds great.

Reply:
I released my "hold", and can review again when you're ready.